### PR TITLE
Support multiple interpolation windows in ForecastTrajectoryGapFiller

### PR DESCRIPTION
**Pull request by gavinevans**
Description
This PR extends the Forecast TrajectoryGapFiller plugin and CLI, so that it is possible to use different interpolation windows for different forecast source pairs. This is desirable because a shorter interpolation window might be desirable at nowcast timescales whilst a longer interpolation window might be desirable at longer lead times.

This PR also fixes a couple of edge cases that were noticed in testing:

1. I noticed that the _identify_periods_to_regenerate wasn't identifying the expected leadtimes for regeneration correctly. Notably although later lead times were being identified fine, the periods being identified were e.g. [3600, 3600, ...]. This was surprising because 3600 (seconds) shouldn't have been identified as a transition, and this transition shouldn't have been included as a duplicate. _identify_source_transitions now groups sources by forecast period first, instead of walking source-by-source. That means if multiple sources share the same period, they are treated as one source state for that period. The transition check is then only between consecutive unique periods, and only when the source set actually changes. That is what removes the spurious 3600 transition and the duplicate 3600 entry. For example, `period_to_sources` variable could look something like below where a transition will only be identified where the set of sources changes.
```python
{
    3600:   {"uk_det", "uk_ens"},
    21600:  {"uk_det", "uk_ens"},
    43200:  {"uk_ens"},
    86400:  {"uk_ens"},
    129600: {"gl_ens"},
    172800: {"gl_ens"},
}
```

2. In the _create_regeneration_tasks method, I noticed that it was possible for regeneration tasks to be silently dropped. This was possible when, for example, the plugin had identified T+1 as a transition lead time, but an interpolation window of 3 hours was being used. This led to the plugin trying to find a T-2 forecast period. As a T-2 forecast period doesn't exist, it was silently dropping this regeneration. There is now support for this situation by a) finding the nearest forecast periods before and after the transition, so that the regeneration of particular lead times won't just be silently dropped, and b) the method is now much noisier if it has to fall back to finding a nearest lead time, so that this sort of edge case won't occur silently.

Testing:

- [x] Ran tests and they passed OK
- [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver/utilities/temporal_interpolation.py [hunk 1]
```diff
@@ -1109,6 +1118,11 @@ def __init__(
             if self.interpolation_window_in_minutes is None
             else self.interpolation_window_in_minutes * 60
         )
+        self.interpolation_window_by_source_pair_seconds = (
+            self._parse_interpolation_window_by_source_pair(
+                interpolation_window_by_source_pair
+            )
+        )
```
Comment (mo-jbeaver) [inline improver/utilities/temporal_interpolation.py:1131]: This section is slightly confusing with the relevant definition below the implementation. It may be helpful to have some comments to explain this section.

File: improver/utilities/temporal_interpolation.py [hunk 2]
```diff
@@ -1460,29 +1642,62 @@ def _create_regeneration_tasks(
...
-                else:
-                    # Fallback: just use the transition point if no interval specified
+            t0_period = expected_t0
+            if expected_t0 not in existing_periods:
+                t0_period = self._find_nearest_period(expected_t0, existing_periods)
+                warnings.warn(
+                    "Regeneration boundary t0 not available for transition "
+                    f"{trans_period}; using nearest period {t0_period} instead of "
+                    f"expected {expected_t0}."
+                )
```
Comment (mo-jbeaver) [inline improver/utilities/temporal_interpolation.py]: The variable is re-named but then both names are used in the following code, this is a bit confusing to understand. This is the same for the section below too.

File: improver/cli/forecast_trajectory_gap_filler.py [hunk 3]
```diff
@@ -74,6 +75,11 @@ def process(
             3 hours and a transition occurs at a given period, periods 3 hours
             before, at, and after the transition will be regenerated if they
             fall within the sequence.
+        interpolation_window_by_source_pair (str):
+            JSON dictionary mapping source-pair keys to interpolation windows in
+            minutes. Keys must identify two source names separated by "|" or
+            "," (for example, '{"uk_ens|gl_ens": 360}'). This overrides the
+            default interpolation_window_in_minutes for matching transitions.
```
Comment (mo-jbeaver) [inline improver/cli/forecast_trajectory_gap_filler.py]:
```suggestion
            "," (for example, '{"uk_ens|gl_ens": 360}'). This overrides the
            default interpolation_window_in_minutes for matching transitions. 
            Matching is order-insensitive. If a source-pair mapping is not
                available for a transition, interpolation_window_in_minutes is used
                as a fallback default if provided.
```
The doctstring from the plugin has a bit more detail, which is helpful.

File: improver_tests/utilities/test_ForecastTrajectoryGapFiller.py [hunk 4]
```diff
@@ -78,11 +80,81 @@ def test_init_custom_parameters():
...
+                388800,
+                432000,
+            ],
+            "uk_det": [3600, 21600],
+        }
+    }
+
+    transitions = plugin._identify_source_transitions(cluster_sources, 0)
+    transition_periods = [period for period, _, _ in transitions]
+    assert transition_periods == [21600, 432000, 648000]
```
Comment (mo-jbeaver) [inline improver_tests/utilities/test_ForecastTrajectoryGapFiller.py:157]: For the different asserts added, it may be helpful to add a comment explaining how these different `transition_periods` and `periods_to_regenerate` are created/determined.

**Review by mo-jbeaver — CHANGES_REQUESTED** [review]
I've added some small comments but, all tests ran successfully.

File: improver/utilities/temporal_interpolation.py (hunk 1 above)
Comment (gavinevans) [reply improver/utilities/temporal_interpolation.py:1131]: I've added a comment to try to make this clearer.

File: improver/utilities/temporal_interpolation.py (hunk 2 above)
Comment (gavinevans) [reply improver/utilities/temporal_interpolation.py]: I've actually split the general section handling warnings into a separate method, so that the purpose can be described a bit clearer and so that it doesn't confuse the surrounding code.

File: improver/cli/forecast_trajectory_gap_filler.py (hunk 3 above)
Comment (gavinevans) [reply improver/cli/forecast_trajectory_gap_filler.py]: This actually prompted me to review some of the fallback behaviour, which I've now simplified, and modified the docstring to reflect this.

File: improver_tests/utilities/test_ForecastTrajectoryGapFiller.py (hunk 4 above)
Comment (gavinevans) [reply improver_tests/utilities/test_ForecastTrajectoryGapFiller.py:157]: I've added comments to the recently added assertions, where there are values asserted, so that it is hopefully a bit clearer where the numbers come from.

**Review by mo-jbeaver — DISMISSED** [review]
Thanks for the additional comments, they were very helpful. The tests all passed successfully, so happy to approve.

File: improver/cli/forecast_trajectory_gap_filler.py [hunk 5]
```diff
@@ -74,6 +75,14 @@ def process(
             3 hours and a transition occurs at a given period, periods 3 hours
             before, at, and after the transition will be regenerated if they
             fall within the sequence.
+        interpolation_window_by_source_pair (str):
+            Optional dictionary mapping forecast source pairs to a transition
+            interpolation window in minutes. Keys must identify two source names,
+            separated by "|" or "," (for example "uk_ens|gl_ens": 360).
+            Matching is order-insensitive. If provided, this takes precedence
+            over interpolation_window_in_minutes and a ValueError is raised if a
```
Comment (brhooper) [inline improver/cli/forecast_trajectory_gap_filler.py:83]: There is no `interpolation_window_in_minutes` argument.

File: improver/utilities/temporal_interpolation.py
Comment (brhooper) [inline improver/utilities/temporal_interpolation.py:1052]: The doc-string suggests this should be `interpolation_window_in_hours`.

File: improver/utilities/temporal_interpolation.py [hunk 6]
```diff
@@ -1061,6 +1062,14 @@ def __init__(
                 forecasts within the interpolation window around a transition will be
                 regenerated, even if they already exist in the input forecast
                 trajectory.
+            interpolation_window_by_source_pair:
+                Optional dictionary mapping forecast source pairs to a transition
+                interpolation window in minutes. Keys must identify two source names,
```
Comment (brhooper) [inline improver/utilities/temporal_interpolation.py:1067]: This will also need updating if the name of `interpolation_window_in_minutes` is changed.

File: improver/utilities/temporal_interpolation.py [hunk 7]
```diff
@@ -1119,6 +1139,120 @@ def __init__(
         self.model_loader = model_loader
         self.kwargs = kwargs
 
+    @staticmethod
+    def _normalise_source_pair_key(key: str) -> FrozenSet[str]:
```
Comment (brhooper) [inline improver/utilities/temporal_interpolation.py]:
```suggestion
    def _prepare_source_pair_key(key: str) -> FrozenSet[str]:
```
I'm not sure what 'normalise' means in this context, I think 'prepare' might be a better verb.

File: improver/utilities/temporal_interpolation.py [hunk 8]
```diff
@@ -1119,6 +1139,120 @@ def __init__(
...
+        """
+        if self.interpolation_window_by_source_pair_seconds:
+            source_pair = frozenset(sources_before | sources_after)
+            if source_pair not in self.interpolation_window_by_source_pair_seconds:
+                available = [
+                    set(p) for p in self.interpolation_window_by_source_pair_seconds
+                ]
+                raise ValueError(
+                    f"No interpolation window configured for source pair "
+                    f"{set(source_pair)}. Available pairs: {available}"
```
Comment (brhooper) [inline improver/utilities/temporal_interpolation.py]: I think calling `set()` here is superfluous as `source_pair` is a frozenset anyway.

File: improver/utilities/temporal_interpolation.py [hunk 9]
```diff
@@ -1272,33 +1406,39 @@ def _identify_source_transitions(
...
+            sources_before = frozenset(period_to_sources[period_before])
+            sources_after = frozenset(period_to_sources[period_after])
 
-            # Only record if source changes
-            if source_before != source_after:
-                # Store the period_before as the transition point.
-                transitions.append(period_before)
+            # Only record if source set changes.
+            if sources_before != sources_after:
+                transitions.append((period_before, sources_before, sources_after))
```
Comment (brhooper) [inline improver/utilities/temporal_interpolation.py]:
```suggestion
        # Find transitions
        transitions = []
        for period_before, period_after in zip(sorted_periods[:-1], sorted_periods[1:]):
            sources_before = frozenset(period_to_sources[period_before])
            sources_after = frozenset(period_to_sources[period_after])

            # Only record if source set changes.
            if sources_before != sources_after:
                transitions.append((period_before, sources_before, sources_after))
```
This is slightly shorter.

File: improver/utilities/temporal_interpolation.py [hunk 10]
```diff
@@ -1337,23 +1480,71 @@ def _identify_periods_to_regenerate(
...
+            minimum_period_exclusive: Lower bound that the returned period must exceed.
+            candidate_periods: Available forecast periods.
+
+        Returns: The nearest available forecast period above the lower bound, or
+            None if no such period exists.
+        """
+        valid_periods = [p for p in candidate_periods if p > minimum_period_exclusive]
+        if not valid_periods:
+            return None
+        return min(valid_periods, key=lambda p: (abs(p - target_period), p))
```
Comment (brhooper) [inline improver/utilities/temporal_interpolation.py]:
```suggestion
    def _find_nearest_period_after(
        self,
        target_period: int,
        minimum_period_exclusive: int,
        candidate_periods: List[int],
    ) -> Optional[int]:
        """Find nearest candidate period above a minimum exclusive bound.

        Args:
            target_period: Forecast period to match.
            minimum_period_exclusive: Lower bound that the returned period must exceed.
            candidate_periods: Available forecast periods.

        Returns: The nearest available forecast period above the lower bound, or
            None if no such period exists.
        """
        valid_periods = [p for p in candidate_periods if p > minimum_period_exclusive]
        if not valid_periods:
            return None
        return self._find_nearest_period(target_period, valid_periods)
```
Since both `_find_nearest_period` and `_find_nearest_period_after` end up performing the same calculation, we could have one call the other to make future maintenance a bit easier.

File: improver/utilities/temporal_interpolation.py [hunk 11]
```diff
@@ -1465,24 +1716,28 @@ def _create_regeneration_tasks(
...
+            if self.interval_in_seconds is not None:
+                # Interpolate only interior periods. Boundary periods are
+                # existing inputs and should not be regenerated.
+                current_period = t0_period + self.interval_in_seconds
+                while current_period < t1_period:
                     interpolation_tasks.append(
-                        ("regenerate", trans_period, expected_t0, expected_t1)
+                        ("regenerate", current_period, t0_period, t1_period)
                     )
+                    current_period += self.interval_in_seconds
```
Comment (brhooper) [inline improver/utilities/temporal_interpolation.py]: `while` loops scare me. Do we have guardrails in place to ensure that `interval_in_seconds` is positive and non-zero?

**Review by brhooper — CHANGES_REQUESTED** [review]
Thanks @gavinevans I've added a few comments.

File: improver/cli/forecast_trajectory_gap_filler.py (hunk 5 above)
Comment (gavinevans) [reply improver/cli/forecast_trajectory_gap_filler.py:83]: I've corrected this docstring.

File: improver/utilities/temporal_interpolation.py
Comment (gavinevans) [reply improver/utilities/temporal_interpolation.py:1052]: I've corrected this docstring.

File: improver/utilities/temporal_interpolation.py (hunk 6 above)
Comment (gavinevans) [reply improver/utilities/temporal_interpolation.py:1067]: I've corrected this docstring.

File: improver/utilities/temporal_interpolation.py (hunk 7 above)
Comment (gavinevans) [reply improver/utilities/temporal_interpolation.py]: Modified.

File: improver/utilities/temporal_interpolation.py (hunk 8 above)
Comment (gavinevans) [reply improver/utilities/temporal_interpolation.py]: Yes, you're right. Removed.

File: improver/utilities/temporal_interpolation.py (hunk 9 above)
Comment (gavinevans) [reply improver/utilities/temporal_interpolation.py]: Adopted.

File: improver/utilities/temporal_interpolation.py (hunk 10 above)
Comment (gavinevans) [reply improver/utilities/temporal_interpolation.py]: Adopted.

File: improver/utilities/temporal_interpolation.py (hunk 11 above)
Comment (gavinevans) [reply improver/utilities/temporal_interpolation.py]: I've replaced this while loop with an equivalent for loop to remove the possibility of infinite looping.

**Review by brhooper — APPROVED** [review]
Thanks for the changes @gavinevans - I'm happy to approve this now.
